Use error prop for violations for consistent styling

trevor-coleman opened the pull request:
### Details

Initially error messages for violations were added as a component under the input. This resulted in odd spacing and behaviour inconsistent with other inputs in the application.

This PR moves the error messages into the `error` prop of the input where it is available.

Because the logic for seelcting which error message to display was getting complicated, I gathered it into a `getErrorForField` function.

I also included the fix for #34773 since it was small and related.

> [!Important]
> Due to an issue outside the scope of this PR, violations will not appear immediately on some fields. 
>
> To get them to update you will need to browse away from the `MoneyRequestView` and back, as that will trigger the necessary update. 
>
> This issue should be solved soon -- it may even be solved before you read this! So if all the expected violations appear immediately in your tests, you can to safely ignore the instructions to browse away and back.

### Fixed Issues

$ https://github.com/Expensify/App/issues/33527
$ https://github.com/Expensify/App/issues/34773

### Tests

#### Create a workspace and enable violations for the policy.

> [!Important] 
> All accounts used in this test need to have the Violations beta enabled in the **back end** in order to receive the correct violations. Ask @cead22 if you need to get accounts added to the beta.

1. Turn on the Violations beta (or set `canUseViolations` to true)
2. Go to **OldDot** and open or create a Control policy
3. Enable violations for the workspace in *Settings > [Workspace] > Expenses*
    1. Expense Violations: true
4. Set up requirements that will a violation on every field:

   | Section    | Settings                                                                                                                                                        |
   |------------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------| 
   | Expenses   | <ul><li>Max Expense Age: 30 days</li> <li>Max Expense Amount: $100</li> <li>Receipt Required Amount: $200</li></ul>                                             | 
   | Categories | <ul><li>People Must Categorize Expenses: ON</li> <li>Set limit for "Advertising" category to $50</li> <li>Set descriptions required for "Advertising"</li></ul> |
   | Tags       | <ul><li>People Must Tag Expenses: ON</li> <li>Add a tag "TestTag"</li></ul>                                                                                     |

5. Enable the policy’s expense chat:
    1. Go to the policy page
    2. In the JS console with the policy open do:

        ```js
        p = Policy.getCurrent();
        p.policy.isPolicyExpenseChatEnabled = "true";
        p.save();
        ```

    3. Invite an **employee** to the policy
    4. Return to New Dot
    5. Sign into the **employee account** you just added to the policy

#### A. First Money Request

1. As the employee go through the money request flow, create the following money request under that policy:

   **Request 1**
    - Receipt - don't include a receipt
    - Amount - $500
    - Category - None
    - Date - 2013-01-01
    - Tag - None
    - Description - leave blank

2. Click another option in the LHN, then browse back to the `MoneyRequestView` and and check that the errors in the table below are shown.

> [!Note]
> A refresh alone will not be enough to get the violations to appear. You will need to browse to another chat and then back to the money request to get the violations to show. (Local processing for violations is being added in a separate PR). 

| Field    | Error                                                                                                       |
|----------|-------------------------------------------------------------------------------------------------------------|
| Receipt  | "Receipt required"                                                                                          |
| Amount   | "Amount greater than $100/person limit"                                                                     |
| Date     | "Date older than 0 days"<br>_(note: "0 days" is a backend bug, but an issue has been opened to address it)_ |
| Tag      | "Missing Tag"                                                                                               |
| Category | "Missing Category                                                                                           |

#### B. Changes to show category violations

1. Update the money to:
    - change the amount to $80
    - change the category to "Advertising"

2. Click another option in the LHN, then browse back to `MoneyRequestView` and check that the errors in the table below are shown. 

> [!Note]
> Violations will not show until you browse away and come back.

> [!Note]
> Errors from previous step will still be visible.


| Field       | Error                                        |
|-------------|----------------------------------------------|
| Amount      | "Amount over C$50.00/person category limit"  |
| Description | "Description required for selected category" |
| Tag         | "Missing Tag"                                |

#### C. Changes to clear violations

1. Update the money to:
    - change the category to "Equipment"
    - Set the tag to "TestTag"
    - Set Date to a date within 30 days

2. Click another option in the LHN, then browse back to the `MoneyRequestView` and check that violation messages are **NOT shown** on `amount`, `category`, `description`, and `tag` fields. 

> [!Note]
> Violations will not be cleared until you browse away and come back.

> [!Note]
> The receipt field will still show a violation.

- [ ] Verify that no errors appear in the JS console

### Offline tests

This will not work offline yet, as we are waiting for local processing of violations to be added in a separate PR.

### QA Steps

- Same as tests above

### PR Author Checklist

- [X] I linked the correct issue in the `### Fixed Issues` section above
- [X] I wrote clear testing steps that cover the changes made in this PR
    - [X] I added steps for local testing in the `Tests` section
    - [X] I added steps for the expected offline behavior in the `Offline steps` section
    - [X] I added steps for Staging and/or Production testing in the `QA steps` section
    - [X] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [X] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [X] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or
      production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using
      the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
        - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer.
          Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using
      the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels
      should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "
      index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working
  as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (
      i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code
  blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component
  like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged
  out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos

<details>
<summary>Android: Native</summary>

A. First Money Request
![android-A](https://github.com/Expensify/App/assets/22041394/440a8704-691f-465c-9ff0-5119d843a3b0)

B. Changes to show category violations
![android-B](https://github.com/Expensify/App/assets/22041394/bea9d17d-8e4f-420b-a74f-511bc6921609)

C. Changes to clear violations
![android-C](https://github.com/Expensify/App/assets/22041394/27fc623a-6ed9-4a9a-b62b-bb5d3ca1021f)

D. Check receipt violations
![android-D](https://github.com/Expensify/App/assets/22041394/b82e3efb-15b7-4abb-9df1-da99564141bb)


</details>

<details>
<summary>Android: mWeb Chrome</summary>

A. First Money Request
![m-web-android-A](https://github.com/Expensify/App/assets/22041394/47887a7b-229b-4926-8fd0-f55d5355aef1)

B. Changes to show category violations
![m-web-android-B](https://github.com/Expensify/App/assets/22041394/bb11903b-b43e-40e8-9cbf-fea182cfe8e6)

C. Changes to clear violations
![m-web-android-C](https://github.com/Expensify/App/assets/22041394/cbe8fd07-4798-46a6-9430-eef479ab925a)

D. Check receipt violations
![m-web-android-D](https://github.com/Expensify/App/assets/22041394/10cedab1-b82d-4f93-b86d-b6fd64f00ec9)


</details>

<details>
<summary>iOS: Native</summary>

A. First Money Request
![ios-A](https://github.com/Expensify/App/assets/22041394/900a5074-40d7-4328-8d89-522b2fefc9d9)

B. Changes to show category violations
![ios-B](https://github.com/Expensify/App/assets/22041394/f15106db-e21c-4c90-a532-518eb7a1e928)

C. Changes to clear violations
![ios-C](https://github.com/Expensify/App/assets/22041394/96818d60-5a96-4214-a931-502d3b01e383)

D. Check receipt violations
![ios-D](https://github.com/Expensify/App/assets/22041394/d46b6a60-1b61-4fa9-86ea-fd1a59b81898)


</details>

<details>
<summary>iOS: mWeb Safari</summary>

A. First Money Request
![m-web-ios-A](https://github.com/Expensify/App/assets/22041394/d045da01-77c7-40c6-aedd-80ea1a2d769b)

B. Changes to show category violations
![m-web-ios-B](https://github.com/Expensify/App/assets/22041394/88f7775b-7966-440b-a98b-ee5935bc4baa)

C. Changes to clear violations
![m-web-ios-C](https://github.com/Expensify/App/assets/22041394/ae7a81aa-273d-4333-b6d3-05e2af98786e)

D. Check receipt violations
![m-web-ios-D](https://github.com/Expensify/App/assets/22041394/7c201d97-8ffa-45df-91f9-3654ec7a75e4)



</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

A. First Money Request
![web-A](https://github.com/Expensify/App/assets/22041394/f62ec049-620e-45cc-a5f5-565cbdc29350)

B. Changes to show category violations
![web-B](https://github.com/Expensify/App/assets/22041394/c866cfd4-4499-46b3-a048-0a8392e24240)

C. Changes to clear violations
![web-C](https://github.com/Expensify/App/assets/22041394/33252d84-5789-483b-acf5-8500529ddcad)

D. Check receipt violations
![web-D](https://github.com/Expensify/App/assets/22041394/18a8824b-ab4c-4422-9f8d-de9c6aecb696)



</details>

<details>
<summary>MacOS: Desktop</summary>

A. First Money Request
![dasktop-A](https://github.com/Expensify/App/assets/22041394/6267180a-e6f9-4ca6-ad3f-3d466a549d95)

B. Changes to show category violations
![desktop-B](https://github.com/Expensify/App/assets/22041394/a12e48d8-50ce-4df6-b9e3-78d73308a445)

C. Changes to clear violations
![desktop-C](https://github.com/Expensify/App/assets/22041394/3015382f-1b04-4739-a692-6c688ccaa951)

D. Check receipt violations
![desktop-D](https://github.com/Expensify/App/assets/22041394/fe6b38b7-30d9-4cff-bb4b-44650e067857)

</details>

trevor-coleman committed:
feat(Violations): create useMoneyRequestViewErrors and move violations into errors prop of inputs

trevor-coleman committed:
Merge remote-tracking branch 'upstream/main' into 32594-update-style-for-violations

trevor-coleman committed:
feat(Violations): move hasViolations into useMoneyRequestViewErrors, and update brickRoadIndicatorProps

trevor-coleman committed:
feat(Violations): add comment

trevor-coleman committed:
feat(Violations): add comment

trevor-coleman committed:
feat(Violations): move getErrorForField into MoneyRequestView

trevor-coleman committed:
Merge branch 'cdanwards/violations/transaction-thread-violations' into 32594-update-style-for-violations

trevor-coleman committed:
feat(Violations): fix import

trevor-coleman committed:
feat(Violations): update translation params to match violation data object

trevor-coleman committed:
feat(Violations): change param names to match violation data prop

trevor-coleman committed:
feat(Violations): change param names to match violation data prop

trevor-coleman committed:
feat(Violations): fixes

trevor-coleman committed:
Merge remote-tracking branch 'upstream/main' into 32594-update-style-for-violations

# Conflicts:
#	src/components/ReportActionItem/MoneyRequestView.js
#	src/libs/ReportUtils.ts
#	src/libs/SidebarUtils.ts
#	src/pages/home/sidebar/SidebarLinksData.js

trevor-coleman committed:
feat(Violations): filter violations by type

trevor-coleman committed:
Merge remote-tracking branch 'upstream/main' into 32594-update-style-for-violations

# Conflicts:
#	src/libs/ReportUtils.ts
#	src/libs/SidebarUtils.ts

trevor-coleman committed:
feat(Violations): fix dependency array

trevor-coleman committed:
feat(Violations): handle default case for missing tag

trevor-coleman commented:
@cead22 -- This is ready to go!

cubuspl42 commented on src/components/ReportActionItem/MoneyRequestView.js:
```diff
@@ -228,6 +229,43 @@ function MoneyRequestView({report, parentReport, parentReportActions, policyCate
+                return ViolationsUtils.getViolationTranslation(violations[0], translate);
+            }
+
+            return '';
```
Do we want to add new code treating `''` as "not set"?

trevor-coleman replied:
I've kept it how it was before, which I think is the sensible choice given that we're in javascript and not typescript, so we don't hav guarantees about how other code calls the component. 

The empty string is falsy, so it'll behave the same way as `null` or `undefined` in any boolean checks, and this will preserve that behaviour.

Previously the empty string was used to mean "not set" as in this example
```ts
error={hasErrors && transactionAmount === 0 ? translate('common.error.enterAmount') : ''}
```


And when it's rendered in `<MenuItem>`, the double shebang `!!('')` essentially casts the value to a boolean so it will gracefully handle booleans, strings, `null` and `undefined`.
```tsx
  {!!errorText && (
   <FormHelpMessage
      isError
      shouldShowRedDotIndicator={false}
      message={errorText}      
      style={styles.menuItemError}
    />
)}
```

trevor-coleman committed:
Merge remote-tracking branch 'upstream/main' into 32594-update-style-for-violations

trevor-coleman commented:
Merged in latest main and it seems to have fixed the failing performance test.

cubuspl42 commented on src/components/ReportActionItem/MoneyRequestView.js:
```diff
@@ -228,6 +229,43 @@ function MoneyRequestView({report, parentReport, parentReportActions, policyCate
+                },
+            };
+
+            const {isError, translationPath} = fieldChecks[field] ? fieldChecks[field] : {};
```
```ts
fieldChecks[field] || {};
```

I think that silencing the linter is better than working it around.

trevor-coleman replied:
done!

cubuspl42 commented:
> Enable the policy’s expense chat ([reference link](https://expensify.slack.com/archives/C05EQBQG6E5/p1695182901255769)):

I don't have access to that link.

> Go to the policy page

It's not clear when I should switch to NewDot.

I can't find my Control policy (workspace) in the NewDot search, and I can't see the main workspace chat in the NewDot LHN. I'm not sure what is supposed to work.

cead22 commented:
> I don't have access to that link.

That links to a message saying the same that's in here, here's a screenshot of the message for context
<img width="579" alt="image" src="https://github.com/Expensify/App/assets/165133/641fcd49-4efb-4d2c-a697-57c03d449858">

cead22 commented on src/components/ReportActionItem/MoneyRequestView.js:
```diff
@@ -379,13 +414,13 @@ function MoneyRequestView({report, parentReport, parentReportActions, policyCate
                                 isOn={transactionBillable}
                                 onToggle={saveBillable}
                             />
+                            {getErrorForField('comment') && (
```
```suggestion
                            {getErrorForField('billable') && (
```

cead22 commented on src/components/ReportActionItem/MoneyRequestView.js:
```diff
@@ -273,10 +311,9 @@ function MoneyRequestView({report, parentReport, parentReportActions, policyCate
                         onPress={() => Navigation.navigate(ROUTES.EDIT_REQUEST.getRoute(report.reportID, CONST.EDIT_REQUEST_FIELD.AMOUNT))}
-                        brickRoadIndicator={hasViolations('amount') || (hasErrors && transactionAmount === 0) ? CONST.BRICK_ROAD_INDICATOR_STATUS.ERROR : ''}
-                        error={hasErrors && transactionAmount === 0 ? translate('common.error.enterAmount') : ''}
+                        brickRoadIndicator={getErrorForField('comment') ? CONST.BRICK_ROAD_INDICATOR_STATUS.ERROR : ''}
```
```suggestion
                        brickRoadIndicator={getErrorForField('amount') ? CONST.BRICK_ROAD_INDICATOR_STATUS.ERROR : ''}
```

cead22 commented on src/components/ReportActionItem/MoneyRequestView.js:
```diff
@@ -341,9 +376,9 @@ function MoneyRequestView({report, parentReport, parentReportActions, policyCate
                             titleStyle={styles.flex1}
                             onPress={() => Navigation.navigate(ROUTES.EDIT_REQUEST.getRoute(report.reportID, CONST.EDIT_REQUEST_FIELD.CATEGORY))}
-                            brickRoadIndicator={hasViolations('category') ? CONST.BRICK_ROAD_INDICATOR_STATUS.ERROR : ''}
+                            brickRoadIndicator={getErrorForField('comment') ? CONST.BRICK_ROAD_INDICATOR_STATUS.ERROR : ''}
```
```suggestion
                            brickRoadIndicator={getErrorForField('category') ? CONST.BRICK_ROAD_INDICATOR_STATUS.ERROR : ''}
```

trevor-coleman replied:
done

cead22 commented on src/components/ReportActionItem/MoneyRequestView.js:
```diff
@@ -355,9 +390,9 @@ function MoneyRequestView({report, parentReport, parentReportActions, policyCate
                             titleStyle={styles.flex1}
                             onPress={() => Navigation.navigate(ROUTES.EDIT_REQUEST.getRoute(report.reportID, CONST.EDIT_REQUEST_FIELD.TAG))}
-                            brickRoadIndicator={hasViolations('tag') ? CONST.BRICK_ROAD_INDICATOR_STATUS.ERROR : ''}
+                            brickRoadIndicator={getErrorForField('comment') ? CONST.BRICK_ROAD_INDICATOR_STATUS.ERROR : ''}
```
```suggestion
                            brickRoadIndicator={getErrorForField('tag') ? CONST.BRICK_ROAD_INDICATOR_STATUS.ERROR : ''}
```

trevor-coleman replied:
done

cead22 commented on src/components/ReportActionItem/MoneyRequestView.js:
```diff
@@ -228,6 +229,43 @@ function MoneyRequestView({report, parentReport, parentReportActions, policyCate
+
+            const {isError, translationPath} = fieldChecks[field] ? fieldChecks[field] : {};
+
+            // Display form errors when first creating the money request
```
```suggestion
            // Return form errors if there are any
```

cead22 commented on src/components/ReportActionItem/MoneyRequestView.js:
```diff
@@ -228,6 +229,43 @@ function MoneyRequestView({report, parentReport, parentReportActions, policyCate
+                return translate(translationPath);
+            }
+
+            // Show violations if there are any
```
```suggestion
            // Return violations if there are any
```

cead22 replied:
I still like this change, since this function doesn't show anything

cead22 replied:
I still like this change, since this function doesn't show anything

cead22 commented on src/hooks/useViolations.ts:
```diff
@@ -4,12 +4,12 @@ import type {TransactionViolation, ViolationName} from '@src/types/onyx';
  * Names of Fields where violations can occur.
  */
-type ViolationField = 'amount' | 'billable' | 'category' | 'comment' | 'date' | 'merchant' | 'receipt' | 'tag' | 'tax';
+type MoneyRequestField = 'amount' | 'billable' | 'category' | 'comment' | 'date' | 'merchant' | 'receipt' | 'tag' | 'tax';
```
Why this change out of curiosity? Isn't ViolationField more accurate? The comment would need to be updated too, but I kinda like ViolationField better than MoneyRequestField.

Reading the comment below "Map from Violation Names to the field where that violation can occur." and the type of `violationFields`, ViolationField fits better imo

cead22 replied:
Any thoughts on this one?

trevor-coleman replied:
Sure doesn't really bother me.

cead22 commented on src/hooks/useViolations.ts:
```diff
@@ -45,27 +45,28 @@ const violationFields: Record<ViolationName, ViolationField> = {
-            const field = violationFields[violation.name];
-            const existingViolations = violationGroups.get(field) ?? [];
-            violationGroups.set(field, [...existingViolations, violation]);
+        const filteredViolations = violations.filter((v) => v.type === 'violation');
```
```suggestion
        const filteredViolations = violations.filter((violation) => violation.type === 'violation');
```

cead22 replied:
Let's use `violation` here per our style guide https://github.com/Expensify/Style-Guides/blob/main/general.md#abbreviations

trevor-coleman replied:
Done

cead22 commented on src/hooks/useViolations.ts:
```diff
@@ -45,27 +45,28 @@ const violationFields: Record<ViolationName, ViolationField> = {
+        const filteredViolations = violations.filter((v) => v.type === 'violation');
+        const violationGroups = new Map<MoneyRequestField, TransactionViolation[]>();
+        for (const violation of filteredViolations) {
+            if (violation.type === 'violation') {
```
We don't  need this here given we already did it for `filteredViolations`, right?

cead22 commented on src/languages/en.ts:
```diff
@@ -2086,7 +2085,7 @@ export default {
         autoReportedRejectedExpense: ({rejectReason, rejectedBy}: ViolationsAutoReportedRejectedExpenseParams) => `${rejectedBy} rejected this expense with the comment "${rejectReason}"`,
         billableExpense: 'Billable no longer valid',
-        cashExpenseWithNoReceipt: ({amount}: ViolationsCashExpenseWithNoReceiptParams) => `Receipt required over ${amount}`,
+        cashExpenseWithNoReceipt: ({formattedLimit}: ViolationsCashExpenseWithNoReceiptParams = {}) => `Receipt required${formattedLimit ? `over ${formattedLimit}` : ''}`,
```
```suggestion
        cashExpenseWithNoReceipt: ({formattedLimit}: ViolationsCashExpenseWithNoReceiptParams = {}) => `Receipt required${formattedLimit ? ` over ${formattedLimit}` : ''}`,
```

cead22 commented on src/languages/es.ts:
```diff
@@ -2584,17 +2584,18 @@ export default {
         receiptNotSmartScanned: 'Recibo no verificado. Por favor, confirma su exactitud',
-        receiptRequired: ({amount, category}: ViolationsReceiptRequiredParams) => `Recibo obligatorio para importes sobre ${category ? 'el limite de la categoría de ' : ''}${amount}`,
+        receiptRequired: ({formattedLimit, category}: ViolationsReceiptRequiredParams = {}) =>
+            `Recibo obligatorio${formattedLimit ? ` para importes sobre ${category ? 'el limite de la categoría de ' : ''}${formattedLimit}` : ``}`,
```
```suggestion
            `Recibo obligatorio${formattedLimit ? ' para importes sobre ${category ? 'el limite de la categoría de ' : ''}${formattedLimit}` : ''}`,
```

trevor-coleman replied:
the first string needs to be template string:
```
` para importes sobre ${category ? 'el limite de la categoría de ' : ''}${formattedLimit}`
```

 but the empty one doesn't

cead22 commented on src/libs/Violations/ViolationsUtils.ts:
```diff
@@ -104,7 +104,7 @@ const ViolationsUtils = {
                 return translate('violations.billableExpense');
             case 'cashExpenseWithNoReceipt':
-                return translate('violations.cashExpenseWithNoReceipt', {amount: violation.data?.amount ?? ''});
+                return translate('violations.cashExpenseWithNoReceipt', violation.data ?? undefined);
```
Out of curiosity, given data is optional (and can be null), why is it necessary to add the fallback to `undefined`? is it because of the type of the second param of `translate`?

trevor-coleman replied:
The function for this translation is:

```
cashExpenseWithNoReceipt: ({formattedLimit}: ViolationsCashExpenseWithNoReceiptParams = {}) => `Receipt required${formattedLimit ? ` over ${formattedLimit}` : ''}`,
```

Which is pretty verbose, so let's take a simpler version

```
myFunction: ({prop}:ParamType = {}) => prop
```

This function takes a single `ParamType` object as a prop and destructures to get the value of prop.

If the anonymous param object is `undefined` then it uses the default value `{}`, and then gets the value of `{}.prop` which is `undefined`

If we pass `null` as the argument it will be used and the code will try to evaluate `null.prop` and throw an error like `TypeError: Cannot read property 'prop' of null`

So we pass undefined to let the function use the default value, which makes `prop` undefined.

I was going back and forth on where to hide the complexity.  In the end I thought it was better to keep the consistency in the translations file, where all the functions are one-liners.

The alternative would be to have something like:
```
cashExpenseWithNoReceipt: (params: ViolationsCashExpenseWithNoReceiptParams | null) => {
    const {formattedLimit} = params ?? {}
    return `Receipt required${formattedLimit ? ` over ${formattedLimit}` : ''
}
```
But that doesn't match the rest of the functions in the translation files, which generally are one liners without logic in there.

I could also define the ViolationsCashExpenseWithNoReceiptParams to be nullable, but again that would be out of line with the rest of the translations, and my general preference in giant files like the translations and their param types is to just be the same as the other code in that file.

trevor-coleman replied:
This is much cleaner now -- I removed the default case and handle the initialization at the top of the function for all variables at once. 

So now it just looks like this:

`return translate('violations.cashExpenseWithNoReceipt', {formattedLimit});`

cead22 replied:
>If the anonymous param object is undefined then it uses the default value {},

Oh right (cause passing undefined is the same as not passing anything), I didn't connect these dots!

cead22 commented on src/libs/Violations/ViolationsUtils.ts:
```diff
@@ -126,30 +126,27 @@ const ViolationsUtils = {
                 return translate('violations.missingComment');
             case 'missingTag':
-                return translate('violations.missingTag', {tagName: violation.data?.tagName});
+                return translate('violations.missingTag', violation.data ?? undefined);
```
If I understand correctly, by passing `violation.data` here, since the type for this is `type ViolationsMissingTagParams = {tagName?: string};`, object deconstructing happens automatically, so we don't have to explocitly pass `tagName`

But in some cases below, we explicitly pass `formattedLimit`. Can we make these consistent?

trevor-coleman replied:
Refactored this to gather all the destructuring and initialization at the top of the function.

There is still one case that I think we should address which is related to `autoReportedRejectedExpense`:

```
autoReportedRejectedExpense: ({rejectReason, rejectedBy}: ViolationsAutoReportedRejectedExpenseParams) => `${rejectedBy} rejected this expense with the comment "${rejectReason}"`,
```
```
autoReportedRejectedExpense: ({rejectedBy, rejectReason}: ViolationsAutoReportedRejectedExpenseParams) => `${rejectedBy} rechazó la solicitud y comentó "${rejectReason}"`,
```

While it should never happen, it's possible in the code as written for `rejectedBy` and `rejectReason` to be `undefined` -- because we can't guarantee that `violation.data` won't be `null`.

Are there default values that make sense for these?

Like maybe `rejectedBy` could be "The policy manager" by default, and we could make "with the comment"
 conditional 

So the strings would be like:

```
The policy administrator rejected this expense
The policy administrator rejected this expense with the comment "sorry"
John Smith rejected this expense
John Smith rejected this expense with the comment "sorry"
```

cead22 replied:
No not really, so maybe to be safe we can just check if neither exists, and return `''`, and if both exist, we return the string we have in there

trevor-coleman replied:
```ts
            case 'autoReportedRejectedExpense':
                return rejectReason && rejectedBy
                    ? translate('violations.autoReportedRejectedExpense', {
                          rejectedBy,
                          rejectReason,
                      })
                    : '';
```

trevor-coleman committed:
feat(Violations): simplify

trevor-coleman committed:
feat(Violations): fix fields

trevor-coleman committed:
feat(Violations): reword comment

trevor-coleman committed:
feat(Violations): remove redundant check

trevor-coleman committed:
feat(Violations): add space

trevor-coleman committed:
feat(Violations): change empty string to single quotes

trevor-coleman committed:
feat(Violations): make translations consistent

trevor-coleman committed:
Merge remote-tracking branch 'upstream/main' into 32594-update-style-for-violations

# Conflicts:
#	src/languages/es.ts

trevor-coleman commented:
Changes in. Merge conflict resolved.

trevor-coleman committed:
feat(Violations): rename vars

trevor-coleman committed:
feat(Violations): return empty string if variables are missing

trevor-coleman commented:
Revisions are in!

cubuspl42 commented on src/languages/es.ts:
```diff
@@ -2584,17 +2584,19 @@ export default {
-        overLimitAttendee: ({amount}: ViolationsOverLimitParams) => `Importe supera el límite de ${amount}/persona`,
-        perDayLimit: ({limit}: ViolationsPerDayLimitParams) => `Importe supera el límite diario de la categoría de ${limit}/persona`,
+        overAutoApprovalLimit: ({formattedLimit}: ViolationsOverAutoApprovalLimitParams) =>
+            `Importe supera el límite de aprobación automática${formattedLimit ? ` de ${formattedLimit}` : ''}`,
+        overCategoryLimit: ({formattedLimit}: ViolationsOverCategoryLimitParams) => `Importe supera el límite para la categoría${formattedLimit ? ` de ${formattedLimit}/persona` : ''}`,
+        overLimit: ({formattedLimit}: ViolationsOverLimitParams) => `Importe supera el límite${formattedLimit ? ` de ${formattedLimit}/persona` : ''}`,
+        overLimitAttendee: ({formattedLimit}: ViolationsOverLimitParams) => `Importe supera el límite${formattedLimit ? ` de ${formattedLimit}/persona` : ''}`,
+        perDayLimit: ({formattedLimit}: ViolationsPerDayLimitParams) => `Importe supera el límite diario de la categoría${formattedLimit ? ` de ${formattedLimit}/persona` : ''}`,
         receiptNotSmartScanned: 'Recibo no verificado. Por favor, confirma su exactitud',
-        receiptRequired: ({amount, category}: ViolationsReceiptRequiredParams) => `Recibo obligatorio para importes sobre ${category ? 'el limite de la categoría de ' : ''}${amount}`,
+        receiptRequired: ({formattedLimit, category}: ViolationsReceiptRequiredParams = {}) =>
+            `Recibo obligatorio${formattedLimit ? ` para importes sobre ${category ? 'el limite de la categoría de ' : ''}${formattedLimit}` : ''}`,
```
Are all the translations final and approved?

trevor-coleman replied:
Yes.

cubuspl42 commented:
Uh, Reassure Performance Tests are failing again 🙁

cubuspl42 commented on src/languages/en.ts:
```diff
@@ -2097,17 +2096,18 @@ export default {
         missingCategory: 'Missing category',
         missingComment: 'Description required for selected category',
-        missingTag: ({tagName}: ViolationsMissingTagParams) => `Missing ${tagName ?? 'tag'}`,
+        missingTag: ({tagName}: ViolationsMissingTagParams = {}) => `Missing ${tagName ?? 'tag'}`,
```
Something is wrong here

trevor-coleman replied:
Can you elaborate on this? Are you experiencing an error?

cubuspl42 replied:
Uhh, I thought for a moment that we're converting an object to a string, although this object is being destructured. Sorry.

trevor-coleman replied:
Ah ok! Phew. 😅

cubuspl42 commented on src/languages/en.ts:
```diff
@@ -2097,17 +2096,18 @@ export default {
-        overLimit: ({amount}: ViolationsOverLimitParams) => `Amount over ${amount}/person limit`,
-        overLimitAttendee: ({amount}: ViolationsOverLimitParams) => `Amount over ${amount}/person limit`,
-        perDayLimit: ({limit}: ViolationsPerDayLimitParams) => `Amount over daily ${limit}/person category limit`,
+        overAutoApprovalLimit: ({formattedLimit}: ViolationsOverLimitParams) => `Expense exceeds auto approval limit of ${formattedLimit}`,
```
Something is wrong here too. Go through all similar places, please.

cubuspl42 commented:
OldDot config:

![image](https://github.com/Expensify/App/assets/2590174/544a31dd-7ca0-4cef-baab-c81447d5293a)

NewDot violations:

![image](https://github.com/Expensify/App/assets/2590174/be7103a4-9c50-468b-b7a6-f106f0d4da58)

(no "Receipt required", "Amount over...")

trevor-coleman commented:
> **A. First Money Request**
> 
> OldDot config:
> 
> ![image](https://private-user-images.githubusercontent.com/2590174/299965538-544a31dd-7ca0-4cef-baab-c81447d5293a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDY1NTMzMDAsIm5iZiI6MTcwNjU1MzAwMCwicGF0aCI6Ii8yNTkwMTc0LzI5OTk2NTUzOC01NDRhMzFkZC03Y2EwLTRjZWYtYmFhYi1jODE0NDdkNTI5M2EucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDEyOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAxMjlUMTgzMDAwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YzliNzI3MzYyODBlMzdiOGEwOGJiZTg5OWQyYjRjOWRlYjE2OTMwNTJlNWU5Njg4NGZlZmFlNTIyNWQ1ZDFlZSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.lCfjfTsS66UGg66LwRRd_lLiGhrl_3U-xSezszbxe34)
> 
> NewDot violations:
> 
> ![image](https://private-user-images.githubusercontent.com/2590174/299965643-be7103a4-9c50-468b-b7a6-f106f0d4da58.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDY1NTMzMDAsIm5iZiI6MTcwNjU1MzAwMCwicGF0aCI6Ii8yNTkwMTc0LzI5OTk2NTY0My1iZTcxMDNhNC05YzUwLTQ2OGItYjdhNi1mMTA2ZjBkNGRhNTgucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDEyOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAxMjlUMTgzMDAwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MjQzOWY4YjI5Mzc3YWNiMjlhZTQxYWIyYmVlM2U5ODhkNDM3ZGU2MTdjNGNlMDczMGY5NTc4YWY3NzI2MmY1ZiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.13ZD1DiqUP7Q6hGdCu4PNpv3cltr6WVYq8hnLuJOgBk)
> 
> (no "Receipt required", "Amount over...")

Can you post a video or steps to reproduce? It's working for me when I create a new violation, but it's possible I've missed something in the instructions so if you can share what you did, or let me know if there's something in my video that I missed in the instructions that'll help me identify the issue.

Video: Adding new money request:
https://github.com/Expensify/App/assets/22041394/42a0381e-76e3-44be-97cf-e07872b19253

And my old dot workspace settings:
![CleanShot 2024-01-29 at 14 07 29@2x](https://github.com/Expensify/App/assets/22041394/317a2466-6912-4fc4-9d19-9096a73338d0)

cubuspl42 commented:
I will post a video! I hope I did everything right.

cubuspl42 commented:
https://github.com/Expensify/App/assets/2590174/8ec0140d-b38e-4ff1-90bb-78c75b58f156

trevor-coleman commented:
Ah yeah  -- so the local state for violations isn't updated until you browse away from the MoneyRequestView and back.  Just refreshing isn't enough because the violations data isn't refetched from the server until you leave and come back.

From the end of your video, try:
1. clicking on a different chat in the LHN (I find its fastest to click on the parent chat of the money request)
2. browse back to the money request, which should now show the correct violations.

You'll need to do this wherever you see the green "tip" callouts in the instructions like this:

> > [!Tip]
> > You will need to browse to another chat and then back to the money request to get
> > the violations to show. (Local processing for violations is being added in a separate PR)

or this:

> > [!Tip]
> > Browse away and back to get the violations to show.

It's a bit annoying, but will hopefully be resolved when the optimistic updates for violations are added -- or they may need to be added on the back end. In any case, the change is outside of the scope of this PR which is just to display any messages that exist in the onyx store in the correct spot.

Looking now I think maybe putting this instruction as a 'tip" callout makes it seem optional, but it's really part of the instructions, so I'll update them to reflect that.

trevor-coleman commented:
updated the test instructions to make it clearer that browsing away and back is an important part of the test.

1. I included the step as a numbered step rather than a callout
2. I moved the callouts to after the instructions, and upgraded them from `[!Tip]` to `[!Note]`
3. Included a note about refreshing not being sufficient.

I think having the instruction in a `[!Tip]` callout made it easy to miss as tips are usually things that are ok to skip, but I think this makes it clearer what's required.

trevor-coleman committed:
Merge remote-tracking branch 'upstream/main' into 32594-update-style-for-violations

trevor-coleman commented:
Merged in main to see if it resovles the performance test.

cead22 commented:
I already have an issue to update `UpdateMoneyRequestAmountAndCurrency` and the other api commands that are similar to that so they push violations to the client

trevor-coleman commented:
Ok great then this is ready for review again. :D

cubuspl42 commented:
Weird, I was sure I switched chats in the meantime... But now I can see the missing violations!

![image](https://github.com/Expensify/App/assets/2590174/fc7f3b69-5f73-4cac-bb8d-94e4f6f995fb)

**But** the receipt placeholder disappeared

![image](https://github.com/Expensify/App/assets/2590174/84d5efa4-a12c-4d88-bbad-c42f0545111f)

Is it expected? The "receipt required" text looks like it's without context.

Also, I can't see the "expense chat" in NewDot in the LHN anymore. I tried re-executing the JS snippet on the OldDot side and refreshing the NewDot tab; didn't help. I'm talking about the workspace chat titled "Control 1" which used to be there.

![image](https://github.com/Expensify/App/assets/2590174/300058c0-5633-45b7-98d0-60d50b9bb886)

Not searchable either.

![image](https://github.com/Expensify/App/assets/2590174/6567b6f6-e1db-4d40-ab21-620475f148b4)

Also, there's a red dot for everything **except** the receipt. But it might be expected.

![image](https://github.com/Expensify/App/assets/2590174/6360ef85-e98b-4741-a44c-eb08820446c0)

cead22 commented:
> Is it expected? The "receipt required" text looks like it's without context.

I think you found a 🐛 , that's not expected

trevor-coleman commented:
Yeah the receipt thing is definitely weird -- will take a look.

The LHN issue is known and being addressed elsewhere. Right now the transaction thread only appears in the LHN while it is open, and then it disappears when you browse away.

trevor-coleman commented:
When the receipt disappears, can you check that you are signed in as the employee who initiated the money request? 

If you are signed in as the admin, I think the receipt placeholder isn't shown when there is no receipt.

The note about it seeming out of context is a good one -- maybe we should show a skeleton or something? This would be a question for the designers, I think.

cead22 commented:
> The note about it seeming out of context is a good one -- maybe we should show a skeleton or something? This would be a question for the designers, I think.

That sounds like a good tweak to me, but out of scope for this PR, so let's not block on it

cubuspl42 commented:
Oh, I accidentally tested with the admin account this time 🙈

So, the thing is that _still_ no amount violation shows to me on the employee side. I switched to different chats multiple times. 

Admin side:

![image](https://github.com/Expensify/App/assets/2590174/ef7888aa-e434-4b1e-8a1a-30e9f408ff74)

Employee side (after switching chats multiple times):

![image](https://github.com/Expensify/App/assets/2590174/cb216e9c-4bf7-4e15-bc4e-0f06999bced7)

cubuspl42 commented:
I always merge-in `main` locally before testing. There's a slight chance that that affected the results. You can merge `main`; it won't hurt.

trevor-coleman committed:
Merge remote-tracking branch 'upstream/main' into 32594-update-style-for-violations

trevor-coleman commented:
Hrm, merging main doens't seem to affect it here. Could you send me a video again from creating the request on? I'm not able to reproduce it here, but maybe there's something I'm missing.

trevor-coleman commented:
Did you create that money request  with the employee account?

Violations should only show for the person who created the money request, so if it was created with the admin account, only the admin should see the violations, but then that means the tag and category violations shouldn't show up.

trevor-coleman committed:
Merge remote-tracking branch 'upstream/main' into 32594-update-style-for-violations

trevor-coleman committed:
Merge remote-tracking branch 'upstream/main' into 32594-update-style-for-violations

# Conflicts:
#	src/components/ReportActionItem/MoneyRequestView.tsx

trevor-coleman commented:
conflict resolved.

trevor-coleman committed:
feat(Violations): restore type

cubuspl42 commented:
I'm unable to reproduce the issue I had before, meaning that this must have been a glitch in the Universe. Things display correctly.

Admin side:

![image](https://github.com/Expensify/App/assets/2590174/17533b59-05fa-4691-bd12-5fa8dfde5824)

Employee side:

![image](https://github.com/Expensify/App/assets/2590174/3b20fb5c-5655-4d3a-89fc-9541792068f7)

cubuspl42 commented:
I found a minor bug:

![image](https://github.com/Expensify/App/assets/2590174/c72eea98-c1c8-4d47-a66c-11caa4567ca6)

No RBR (red dot) is displayed, which violates Expensify principles. Is it in the scope of this PR?

cubuspl42 commented:
> Could you send me a video again from creating the request on?

https://github.com/Expensify/App/assets/2590174/7c65d457-76a9-4957-92e7-4462dd1cfe49

This is weird... It seems to work for one money request I made (the $567 one), but failed again for another request (the $789 one).

trevor-coleman commented:
@cubuspl42  -- I think I know what's happening -- in that video you browse away from the transaction thread to somewhere else and then back to the transaction thread directly. Can you try browsing away from the transaction thread, then open the parent policy chat and then browse to the money request from there?

If I'm right that should trigger the update.

@carlos -- can you look at the video above? I think what's happening is that the MoneyRequestView in the transaction thread is not being updated with the latest onyx data.It's only when the parent report is opened that the database is updated.

trevor-coleman committed:
Merge remote-tracking branch 'upstream/main' into 32594-update-style-for-violations

# Conflicts:
#	src/components/ReportActionItem/MoneyRequestView.tsx

trevor-coleman commented:
Resovled conflicts

cead22 commented:
Conflicts are back 😅

cubuspl42 commented:
Linter & conflicts

cubuspl42 commented:
@trevor-coleman So I understand that we're fine with the _partial_ validation being observable, assuming that I manage to force a "full refresh" during testing

cubuspl42 commented:
@trevor-coleman Bump on: https://github.com/Expensify/App/pull/34303#issuecomment-1927002060

cubuspl42 commented:
> If I'm right that should trigger the update.

I'm really trying 🙁

https://github.com/Expensify/App/assets/2590174/c28357c1-8611-41ef-8bb3-0b58320b1ead

It's kind of getting worse. Now I see a red dot near "Tag", but no error message.

trevor-coleman committed:
Merge remote-tracking branch 'origin/main' into 32594-update-style-for-violations

# Conflicts:
#	src/libs/Violations/ViolationsUtils.ts

trevor-coleman committed:
feat(Violations): add deps to array

trevor-coleman commented:
In that one you are signed in as David Oppenheim who is the admin for the policy, aren't you?
![CleanShot 2024-02-06 at 11 49 57@2x](https://github.com/Expensify/App/assets/22041394/05fbf7de-650c-4b19-9142-7334c2d1a5ef)

![CleanShot 2024-02-06 at 11 50 31@2x](https://github.com/Expensify/App/assets/22041394/16599121-2179-45fa-8861-3642fe77959c)

![CleanShot 2024-02-06 at 11 51 17@2x](https://github.com/Expensify/App/assets/22041394/caa86ae9-0c09-4b5a-8e25-79808259ef3c)

trevor-coleman commented:
The linter error is from `main` on [this commit](https://github.com/Expensify/App/commit/13f832ca22b3d6802f60d0031bc782ed5185d719) -- I'll merge in the fix whenever we get it.

trevor-coleman committed:
feat(Violations): replace tag field

trevor-coleman committed:
Merge remote-tracking branch 'upstream/main' into 32594-update-style-for-violations

trevor-coleman commented:
Merged main again, linter error should be gone.

cubuspl42 commented:
> In that one you are signed in as David Oppenheim who is the admin for the policy, aren't you?

No! The auto-generated avatars are mixed up. I don't know why, it's on the Expensify side.

>  as David Oppenheim

I just refer to them by numbers 😅 These are 8 and 9. I changed the avatars now, so it's not confusing anymore.

![image](https://github.com/Expensify/App/assets/2590174/4591da77-f594-45e0-a354-3aed16a555c7)

I re-tested again:

https://github.com/Expensify/App/assets/2590174/2c4e17d7-c0f9-4d51-a61b-ef0c12558f14


I'll remind that I do have the violations beta enabled:

```ts
function canUseViolations(betas: OnyxEntry<Beta[]>): boolean {
    return true;
    return !!betas?.includes(CONST.BETAS.VIOLATIONS) || canUseAllBetas(betas);
}

```

cubuspl42 commented:
Also, I'll note that I have one money request that I **can** consistently observe the violations in. It's the $567 one. It's the first one in the group (in a transaction?). Maybe that's the reason? Is the new feature supposed to work only for that money request?

https://github.com/Expensify/App/assets/2590174/33a3f236-8672-4a37-918d-cbe9add676ed

cubuspl42 commented:
But I also see the violations in the $567 request from the admin side (account 8), and you say I shouldn't =(

![image](https://github.com/Expensify/App/assets/2590174/288210b7-a821-4514-bb56-464f8e679682)

trevor-coleman commented:
I wonder if the "+" in the email is causing confusion between the accounts. Like maybe there's a bad regex somewhere.

trevor-coleman commented:
Another question: Has that report been submitted? 

If it's submitted, then the admin will see the violations and the requester will not see them.

trevor-coleman commented:
Ok I asked Carlos to look into it and the `+8` account did not have the violations beta enabled in the back end. 

He's enabled it now, so give it another shot and see if anything has changed. 

I'll also add somethig to the instructions to make sure that both accounts have the beta enabled in the back end as well as in the front end.

trevor-coleman commented:
Added this block to help future testers:

> [!Important] 
> All accounts used in this test need to have the Violations beta enabled in the **back end** in order to receive the correct violations. Ask @cead22 if you need to get accounts added to the beta.

trevor-coleman commented:
The +7 and +9 emails have also been added.

trevor-coleman commented:
I'm not sure what is going on, tbh. I wonder if something got messed up with the policy data at some point with the non-beta accounts working with the violations data. 

I've not been able to reproduce it here and I had a teammember run through the instructions and it worked. Can you maybe try with a completely fresh start -- clear the local onyx db and start a fresh policy?

cead22 commented:
I think it's possible everything is working correctly, but if you can share the workspace IDs you're using, and maybe more specifically what you think it's not working, I can confirm.

For the things that look strange there
- We have a feature that works in the back end called auto categorization or implicit categorization, and in short if you set Advertising as the category for a request with the merchant set to A, the next time you create a request with a merchant set to A and don't categorize it, we'll _infer_ that the category should be Advertising
- You don't see this for other requests, eg the one with merchant set to B, because it's not the same merchant as A
- Submitting is separate from violations so I don't know if there's anything wrong there, but even unsubmitted requests should show violations, so that seems to be working fine

cead22 commented:
Feel free to ask any questions and I'll be happy to add clarity where I can, I know there's a lot of moving parts and some non-obvious things like implicit categorization that makes things confusing

cubuspl42 commented:
A.

https://github.com/Expensify/App/assets/2590174/90ff2a35-2a9e-46fa-81db-afc400d6b066

B.

https://github.com/Expensify/App/assets/2590174/01f10099-8546-4161-bb55-c94d3a83b557

C.

https://github.com/Expensify/App/assets/2590174/a266fc43-58c4-4983-afe2-c01dfdd4ba47

D.

https://github.com/Expensify/App/assets/2590174/0bc615e1-c187-4f1f-878e-99acc5696ac9

![walmart_receipt](https://github.com/Expensify/App/assets/2590174/7941030d-e411-428a-96b3-a0dd7960efa2)

----

Would you take a look at the test D? I couldn't get the "Amount is greater than scanned receipt" text to show up.

trevor-coleman commented:
Awesome! Glad we got all that sorted. Let me look into the scanning one.

cead22 commented:
I think that's happening because we're not showing that violation (by design) on money request while the receipt is scanning. We only show it after the receipt finishes scanning. 

Can you check again to see if it's showing now? I checked the database and it looks like SmartScan completed, and looking at the api response for a command to load violations, I see the violation in there now

<img width="538" alt="image" src="https://github.com/Expensify/App/assets/165133/29cd40a1-6bbe-411e-9685-af491bbbecc4">

cubuspl42 commented:
> while the receipt is scanning

It can be seen on the video that the receipt has finished scanning from the user's perspective. The price 23.19 comes from the receipt (although the currency got messed up).

<img width="180" alt="image" src="https://github.com/Expensify/App/assets/2590174/28f14375-9aa1-4bfe-a8f3-04bfc794e830">

cead22 commented:
> It can be seen on the video that the receipt has finished scanning from the user's perspective

Yes, when users input the data being scanned manually, the request looks like it's not scanning, but I believe we continue the scan in the background.

Is the violation showing now?

> The price 23.19 comes from the receipt (although the currency got messed up).

Can you share more details about this, are you pointing to a specific error or problem with the PR? Sorry if I'm missing something obvious, I'm just trying to figure out if things are working correctly, and so far it seems like they are

trevor-coleman commented:
I think I've figured it out -- the `modifiedAmount` violation is a `notice` type violation, so it won't be displayed. 

That `notice` type filter was part of a separate issue that I brought into this PR after the test instructions were written, and I didn't realize `modifiedAmount` was that kind of violation.

I've removed section D from the test instructions, and I think we are now passing all tests. 

Just need to merge in main to fix the lint error and we are good.

trevor-coleman committed:
Merge remote-tracking branch 'upstream/main' into 32594-update-style-for-violations

cead22 commented:
Awesome sauce!

trevor-coleman commented:
Main is merged, and now comes the time of anxious waiting for the CI checks to complete.

cead22 commented:
Am I crazy or is the check for the reviewer checklist missing?

situchan reviewed (changes requested):
(just for testing checks)
Is reviewer checklist missing

cubuspl42 commented:
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos

<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/2590174/effa9f23-32af-496e-88bf-15d063a3fbd2

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/2590174/696a8ebf-d28a-4ef3-919a-90cc9c40a2f8

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/2590174/ef2fa205-d8ae-4ec0-a631-b00facfee510

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/2590174/0d815e9a-1454-44ba-b514-f484725bc9f1

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

A.

https://github.com/Expensify/App/assets/2590174/90ff2a35-2a9e-46fa-81db-afc400d6b066

B.

https://github.com/Expensify/App/assets/2590174/01f10099-8546-4161-bb55-c94d3a83b557

C.

https://github.com/Expensify/App/assets/2590174/a266fc43-58c4-4983-afe2-c01dfdd4ba47

</details>

<details>
<summary>MacOS: Desktop</summary>

<!-- add screenshots or videos here -->

</details>

trevor-coleman committed:
Merge remote-tracking branch 'upstream/main' into 32594-update-style-for-violations

trevor-coleman committed:
Merge remote-tracking branch 'upstream/main' into 32594-update-style-for-violations

# Conflicts:
#	src/components/ReportActionItem/MoneyRequestView.tsx

trevor-coleman commented:
So close!

cead22 commented:
Yeah I just wanted to test dekstop becasue I didn't see screenshots/videos in the reviewer checklist. I added them now

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/7911416146) to production by https://github.com/thienlnam in version: 1.4.41-12 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|failure ❌
🕸 web 🕸|success ✅
